Add support for a postfix block with sh

jdee opened the pull request:
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

Partly this is motivated by @janpio's [issue](https://github.com/fastlane/fastlane/issues/11140) and partly by spending quality time with Rake.

Working with Procs can be awkward. To have a multiline `error_callback`, you have choices like:

```Ruby
callback = Proc.new do |result|
  # multiline error handler
end

sh command, error_callback: callback
```

or

```Ruby
sh(command, error_callback: ->(result) { handle_error(result) }) # call a method
```

Postfix blocks are easier to use syntactically and more idiomatic. Rake supports something like this in its `sh` method. I've often thought it would be nice to allow blocks with actions, but it would require some work in the runner context. However, `sh` is not an action, so this is fairly trivial.

With this change:

```Ruby
sh command do |status, result|
  if status.success?
    UI.success "Command succeeded"
  else
    UI.user_error "Command failed with status #{status.exitstatus}"
  end
end
```

### Description

This just passes blocks from all variants of `sh` down to `Actions.sh_control_output`, which yields if `block_given?`.

This also replaces the empty fixture file in the unit tests with a more general `StringIO` that allows for mocking output.

jdee committed:
Add support for a postfix block with sh

ohayon reviewed (commented):
Hey @jdee - is adding the `&b` param to those methods necessary? Can't ruby handle the block param for any method and it doesnt need to be explicitly listed in the method definition?

jdee commented:
@ohayon I don't think it's necessary in `Actions.sh_control_output`. I don't think it's passed through multiple method calls without passing it as an arg. I can review. Less is more.

jdee commented:
It's necessary to support blocks in all the variants, but it can be removed from `sh_control_output`. Here's a little test code to demonstrate:

```Ruby
def f1(arg)
  f2 arg
end

def f2(arg)
  yield arg if block_given?
end

f2 "foo" do |arg|
  puts "result from f2: #{arg.inspect}"
end

f1 "foo" do |arg|
  puts "result from f1: #{arg.inspect}"
end
```

```bash
[jdee@Jimmy-Dees-MacBookPro tmp]$ ruby test.rb
result from f2: "foo"
```

The block is not invoked when calling `f1`. Then if i change `f1`:

```Ruby
def f1(arg, &b)
  f2 arg, &b
end
```

```bash
[jdee@Jimmy-Dees-MacBookPro tmp]$ ruby test.rb
result from f2: "foo"
result from f1: "foo"
```

I can remove it from `sh_control_output`.

jdee commented:
I was also thinking of using `Forwardable` if possible to remove some of the wrappers methods here and there, like `Action.sh`.

jdee committed:
Use Forwardable module to delegate from Action.sh to Actions.sh_control_output

jdee committed:
Clean-up

jdee committed:
Updated gradle_spec

jdee committed:
Remove unnecessary module name

ohayon commented on fastlane/lib/fastlane/helper/sh_helper.rb:
```diff
@@ -62,14 +70,15 @@ def self.sh_control_output(*command, print_command: true, print_command_output:
         result << shell_command # only for the tests
       end
 
+      return block_return if block_given?
```
Will this still end up reporting the error if it skips that on line 73 if there is a block?

ohayon reviewed (approved):
This looks really awesome @jdee ! Just one question above, but might not require any changes. So go ahead and merge when ready! 🚀

jdee committed:
Be sure to output any error to the user even when a block is given.

jdee committed:
Too many ifs for RuboCop's taste.

jdee committed:
Also yield the equivalent shell command.

jdee committed:
Notify block after error message is logged.

jdee commented:
@ohayon Thanks for the approval. In the cut that you reviewed, an error message was not directly presented to the user via UI, unlike the other invocation methods. This seems best for consistency, though of course the block will get sufficient information to show the same information.

I made that change and then also moved the yield until after the error is logged, or you'll get the command output after a custom error message from the block.

Finally, I also added the constructed shell command to the yield, for ease of logging when using a gnarly Ruby expression for the argument. So now you can easily do this:

```Ruby
  sh "ls", "-la", "foobar" do |status, result, command|
    UI.error %(Command #{command} (pid #{status.pid}) failed with status #{status.exitstatus}) unless status.success?
    UI.message "Result is #{result.inspect}"
  end
```

and generate output like

```
[09:01:07]: ---------------------------
[09:01:07]: --- Step: ls -la foobar ---
[09:01:07]: ---------------------------
[09:01:07]: $ ls -la foobar
[09:01:07]: ▸ ls: foobar: No such file or directory
[09:01:07]: Exit status of command 'ls -la foobar' was 1 instead of 0.
ls: foobar: No such file or directory

[09:01:07]: Command ls -la foobar (pid 94338) failed with status 1
[09:01:07]: Result is "ls: foobar: No such file or directory\n"
```

The last two lines come from the block.

I'll merge this later, but if you have any further thoughts, please weight in.

ohayon commented:
No more thoughts this is 💯 !
